rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,1222 @@
+each teleport instance in need of upgrade to run a user-supplied script locally and
+then restart). From the perspective of a user, an installer is a teleport
+configuration object (though that configuration object may only be a thin hook into
+the "real" `install-controller`). Whether or not the teleport instance being upgraded
+needs to understand the installer will vary depending on type.
+
+There is room in the above model for additional granularity (e.g. schedulers), but
+the listed components are a decent starting point.
+
+
+### Implementation Phases
+
+Implementation will be divided into a series of phases consisting of 1 or more
+separate PRs/releases each:
+
+- Setup: Changes required for inventory status/control model, but not necessarily
+specific to the upgrade system.
+
+- Notification-Only System: Optional phase intended to deliver value sooner
+at the cost of overall feature development time.
+
+- Script-Based Installs MVP: Early MVP supporting only manual control and
+simple script-based installers for non-auth instances.
+
+- TUF-Based System MVP: Fully functional but minimalistic upgrade system
+based on TUF (still excludes auth instances).
+
+- Stability & Polish: Additional necessary features (including auth installs
+and local rollbacks). Represents the point at which the core upgrade system
+can be considered "complete".
+
+- Extended Feature Set: A collection of nice-to-haves that we're pretty sure
+folks are going to want.
+
+See the [implementation plan](#implementation-plan) section for detailed breakdown
+of the required elements for each phase.
+
+
+## Details
+
+### Usage Scenarios
+
+
+#### Notification-Only Usecase
+
+Cluster administrators may or may not be using teleport's own install mechanisms, but they
+want teleport to be able to inform them when instances are outdated, and possibly generate
+alerts if teleport is running a deprecated version and/or there is a newer security patch
+available.
+
+In this case we want options for both displaying suggested version (or just a "needs upgrade"
+badge) on inventory views, and also probably some kind of cluster-wide alert that can be
+made difficult to ignore (e.g. a message on login or a banner in the web UI). We also
+probably want an API that will support plugins that can emit alerts to external locations
+(e.g. a slack channel).
+
+In this usecase teleport is serving up recommendations based on external state, so a client
+capable of discovery (e.g. the `tuf` version controller) is required, but the actual
+ability to affect installations may not be necessary.
+
+
+#### Minimal Installs Usecase
+
+Cluster administrators explicitly set the desired version(s) of teleport installations, and
+have a specific in-house installation script that should be used for upgrades. Teleport cluster
+may not even have any access to the public internet.  The installation target may be an internally
+built fork of teleport.
```

Comment:
Supporting installation of forks may be an important feature, but I'm not sure I would describe it as a "minimal" use case. There's lots to consider there. Is this feature something we can defer?

Reply:
Its minimal in the sense that teleport's responsibilities/involvement is minimal.  This scenario is about the case where users are just using teleport to trigger their own install mechanism that is totally unspecified from teleport's perspective. The user-supplied script does everything.  The script might install a fork... it also might just create a file named `plz-upgrade-soon.txt` on the host that the interns know to look for.

Reply:
> There are only two hard things in Computer Science: cache invalidation and naming things.

Any suggestions?  I've become accustomed to us using terms like client/server when referring to the relationship between auth and non-auth instances, but I'm good with anything that evokes directionality.